packages/cache/src/internal/cacheHttpClient.ts

```diff
@@ -0,0 +1,339 @@
+      throw new Error(
+        `Incomplete download. Expected file size: ${expectedLength}, actual file size: ${actualLength}`
+      )
+    }
+  } else {
+    core.debug('Unable to validate download, no Content-Length header')
+  }
+}
+
+// Reserve Cache
+export async function reserveCache(
+  key: string,
+  paths: string[],
+  options?: CacheOptions
+): Promise<number> {
+  const httpClient = createHttpClient()
+  const version = getCacheVersion(paths, options?.compressionMethod)
+
+  const reserveCacheRequest: ReserveCacheRequest = {
+    key,
+    version
+  }
+  const response = await httpClient.postJson<ReserveCacheResponse>(
+    getCacheApiUrl('caches'),
+    reserveCacheRequest
+  )
+  return response?.result?.cacheId ?? -1
+}
+
+function getContentRange(start: number, end: number): string {
+  // Format: `bytes start-end/filesize
+  // start and end are inclusive
+  // filesize can be *
+  // For a 200 byte chunk starting at byte 0:
+  // Content-Range: bytes 0-199/*
+  return `bytes ${start}-${end}/*`
+}
+
+async function uploadChunk(
+  httpClient: HttpClient,
+  resourceUrl: string,
+  data: NodeJS.ReadableStream,
+  start: number,
+  end: number
+): Promise<void> {
+  core.debug(
+    `Uploading chunk of size ${end -
+      start +
+      1} bytes at offset ${start} with content range: ${getContentRange(
+      start,
+      end
+    )}`
+  )
+  const additionalHeaders = {
+    'Content-Type': 'application/octet-stream',
+    'Content-Range': getContentRange(start, end)
+  }
+
+  const uploadChunkRequest = async (): Promise<IHttpClientResponse> => {
+    return await httpClient.sendStream(
+      'PATCH',
+      resourceUrl,
+      data,
+      additionalHeaders
+    )
+  }
+
+  const response = await uploadChunkRequest()
+  if (isSuccessStatusCode(response.message.statusCode)) {
+    return
+  }
+
+  if (isRetryableStatusCode(response.message.statusCode)) {
+    core.debug(
+      `Received ${response.message.statusCode}, retrying chunk at offset ${start}.`
+    )
+    const retryResponse = await uploadChunkRequest()
+    if (isSuccessStatusCode(retryResponse.message.statusCode)) {
+      return
+    }
+  }
+
+  throw new Error(
+    `Cache service responded with ${response.message.statusCode} during chunk upload.`
+  )
+}
+
+function parseEnvNumber(key: string): number | undefined {
+  const value = Number(process.env[key])
+  if (Number.isNaN(value) || value < 0) {
+    return undefined
+  }
+  return value
+}
+
+async function uploadFile(
+  httpClient: HttpClient,
+  cacheId: number,
+  archivePath: string
+): Promise<void> {
+  // Upload Chunks
+  const fileSize = fs.statSync(archivePath).size
+  const resourceUrl = getCacheApiUrl(`caches/${cacheId.toString()}`)
+  const fd = fs.openSync(archivePath, 'r')
+
+  const concurrency = parseEnvNumber('CACHE_UPLOAD_CONCURRENCY') ?? 4 // # of HTTP requests in parallel
+  const MAX_CHUNK_SIZE =
+    parseEnvNumber('CACHE_UPLOAD_CHUNK_SIZE') ?? 32 * 1024 * 1024 // 32 MB Chunks
```

Comment:
Do we want to add upload concurrency and chunk size as optional parameters?

Reply:
Yeah let's make these optional inputs instead of environment variables.

We should do something similar to [`globOptions`](https://github.com/actions/toolkit/blob/master/packages/glob/src/glob.ts#L14) so that we don't have to change the function signature everytime we add a new option.